docs/CIP_6_Pipelines_Registry.md

```diff
@@ -0,0 +1,309 @@
+
+```python
+collection = client.create_collection("my_collection", pipeline=DefaultEmbeddingFunctionPipelineStep())
+
+# ... some time later
+
+collection = client.get_collection("my_collection") # The default pipeline is automatically loaded
+```
+
+#### Query Pipelines
+
+In order to support embedding the data automatically for the user on query, we propose the introduction of the query_pipeline, which is the default pipeline run on the query data. This pipeline is persisted with the collection, and is used when no pipeline is provided. This means that users can create a collection with a default query pipeline, and then query the collection without having to provide the query pipeline.
+
+For example, the default query pipeline would simply Embed the text, and would be defined as follows:
+
+```python
+collection = client.create_collection("my_collection", pipeline=DefaultEmbeddingFunctionPipelineStep(), query_pipeline=DefaultEmbeddingFunctionPipelineStep()) # The defaults would not actually have to be specified, as they would be the default values. Here we are specifying them for clarity.
+collection.add(ids, data=["text_to_embed"]) # The content of the data is embedding using the pipeline and then added to the collection
+collection.query(data=["text_to_query"]) # The content of the data is embedding using the query pipeline and then queried against the collection
+```
+
+This allows for the pipelines at insert time, and the pipelines at query time to be different. This is useful for a variety of reasons. For example, if you wanted to chunk the data at insert time, but not at query time, you could do something like this:
+
+```python
+collection = client.create_collection("my_collection", pipeline=Embed(Chunk()), query_pipeline=Embed())
+```
+
+#### The I/0 of a pipeline stage
+The input to a pipeline step is the set of parameters provided to an insert call in Chroma - ids, documents, metadata and embeddings. The size of each of these fields must match or be empty. This way, pipeline stages can both pass through data to subsequent stages, or they can populate them themselves. The output of the final stage of a pipeline must be a valid input to the insert call in Chroma, i.e it must contain ids, documents, metadata and embeddings. For example, imagine a pipeline that consumes documents, chunks them, and then embeds them.
+
+```python
+data = ["large_document_1", "large_document_2"]
+pipeline = Embed(Chunk())
+collection = client.create_collection("my_collection", pipeline=pipeline)
+collection.add(data=data, ids=[0, 1])
+
+# Input to pipeline would be {ids: [0, 1], documents: ["large_document_1", "large_document_2"], metadata: None, embeddings: None}
+# Output of Chunk could be {ids: [A, B, C, D], documents: ["chunk_1", "chunk_2", "chunk_3", "chunk_4"], metadata: [{"source_id": 0}, {"source_id": 0}, {"source_id": 1}, {"source_id": 1}], embeddings: None}
```

Comment:
@HammadB, ok I see that my point above is kind of moot with this but still, do you think it could be useful to "evidence" somehow to the user that e.g. metadata will be injected etc - perhaps through function documentation e.g. what are the side effects of the pipeline function.

On the same note, maybe we can have a utility method that will create a local representation of the transformed data so the user can review it prior to storing it in Chroma.

Reply:
Yeah, one general concern I have about this, is the somewhat opaque transformation of data that happens for you. For example if you assign ids, how do those ids get returned to you? How can you see results of pipeline execution?

Reply:
Pipelines need a good introspection, I reckon. But maybe just start with logging :)